exonum/src/events/noise/sodium_resolver.rs

```diff
@@ -0,0 +1,243 @@
+use byteorder::{ByteOrder, LittleEndian};
+use rand::{thread_rng, Rng};
+use snow::params::{CipherChoice, DHChoice, HashChoice};
+use snow::types::{Cipher, Dh, Hash, Random};
+use snow::{CryptoResolver, DefaultResolver};
+
+use sodiumoxide::crypto::aead::chacha20poly1305 as sodium_chacha20poly1305;
+use sodiumoxide::crypto::hash::sha256 as sodium_sha256;
+use sodiumoxide::crypto::scalarmult::curve25519 as sodium_curve25519;
+
+pub struct SodiumResolver {
+    parent: DefaultResolver,
+}
+
+impl SodiumResolver {
+    pub fn new() -> Self {
+        SodiumResolver {
+            parent: DefaultResolver,
+        }
+    }
+}
+
+impl CryptoResolver for SodiumResolver {
+    fn resolve_rng(&self) -> Option<Box<Random + Send>> {
+        Some(Box::new(SodiumRandom::default()))
+    }
+
+    fn resolve_dh(&self, choice: &DHChoice) -> Option<Box<Dh + Send>> {
+        match *choice {
+            DHChoice::Curve25519 => Some(Box::new(SodiumDh25519::default())),
+            _ => self.parent.resolve_dh(choice),
+        }
+    }
+
+    fn resolve_hash(&self, choice: &HashChoice) -> Option<Box<Hash + Send>> {
+        match *choice {
+            HashChoice::SHA256 => Some(Box::new(SodiumSha256::default())),
+            _ => self.parent.resolve_hash(choice),
+        }
+    }
+
+    fn resolve_cipher(&self, choice: &CipherChoice) -> Option<Box<Cipher + Send>> {
+        match *choice {
+            CipherChoice::ChaChaPoly => Some(Box::new(SodiumChaChaPoly::default())),
+            _ => self.parent.resolve_cipher(choice),
+        }
+    }
+}
+
+// Random data generator.
+struct SodiumRandom;
+
+impl Default for SodiumRandom {
+    fn default() -> SodiumRandom {
+        SodiumRandom {}
+    }
+}
+
+impl Random for SodiumRandom {
+    fn fill_bytes(&mut self, out: &mut [u8]) {
+        let bytes: Vec<u8> = thread_rng().gen_iter::<u8>().take(out.len()).collect();
+        out.copy_from_slice(&bytes);
+    }
+}
+
+// Elliptic curve 25519.
+pub struct SodiumDh25519 {
+    privkey: sodium_curve25519::Scalar,
+    pubkey: sodium_curve25519::GroupElement,
```

Comment:
Does `sodiumoxide` has no wrapers around `Curve25519` keys?
Can we implement it in `crypto` module?

Reply:
Those `Scalar` and `GroupElement` are exactly the wrappers for keys. You can find them in `sodiumoxide/crypto/scalarmult/curve25519.rs`.

Reply:
If you mean a `struct` that contain both `Scalar` and `GroupElement`, then no. Then it can be implemented in `crypto`, yep.

Reply:
`Scalar` and `GroupElement` are wrappers around `curve` primitives which in ECDH used as `SecretKey`\`PublicKey`. 
I am talking about abstraction `SecretKey`, `PublicKey` in submodule `x25519` in scope of `crypto.rs` (analog of existing `ed25519`)

Reply:
Got it.